### PR TITLE
Feature/dr 2925/fallback for featured item

### PR DESCRIPTION
**Pull request by avertrees**
## Ticket:

- JIRA ticket [DR-2925](https://jira.nypl.org/browse/DR-2925)

## This PR does the following:

- updates hero component to use a fallback image for the featured item

## How has this been tested?

¯\_(ツ)_/¯ 

## Accessibility concerns or updates

-

### Checklist:

- [ ] I have added relevant accessibility documentation for this pull request.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/types/FeaturedItemData.ts
Comment (avertrees) [inline src/types/FeaturedItemData.ts:1]: I was optimistic in trying to use types for the first time but then I got confused and gave up

File: src/data/defaultFeaturedItemData.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
+const defaultFeaturedItem: any = {
+  development: {
+    featuredItem: {
+      imageID: "482815",
+      backgroundImageSrc: "/482815.jpg",
+      foregroundImageSrc: "/482815.jpg",
+      imageSrc: "/482815.jpg",
```
Comment (EdwinGuzman) [inline src/data/defaultFeaturedItemData.ts:7]: You don't seem to be using the `imageSrc` property anywhere.

File: src/data/defaultFeaturedItemData.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
+const defaultFeaturedItem: any = {
```
Comment (EdwinGuzman) [inline src/data/defaultFeaturedItemData.ts:1]: Here is a case where you can use the `FeaturedItemData` interface.

If you had a single object, you can type it as the following as an example:

```
const development: FeaturedItemData = {
  featuredItem: {
    imageID: "482815",
    backgroundImageSrc: "/482815.jpg",
    foregroundImageSrc: "/482815.jpg",
    uuid: "510d47d9-4f93-a3d9-e040-e00a18064a99",
    title: "Watuppa, From water front, Brooklyn",
    href: "https://qa-digitalcollections.nypl.org/items/510d47d9-4f93-a3d9-e040-e00a18064a99",
  },
  numberOfDigitizedItems: "875,861",
};
```

To be clear, DON'T change your data structure, it's fine but it needs an additional typescript utility type to help define the entire structure. What you can use is the `Record` utility type to help https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeys-type

I'll leave the rest for you to work through but let me know if you run into issues.

File: src/components/hero/campaignHero.tsx [hunk 3]
```diff
@@ -4,9 +4,19 @@ import { useEffect, useState } from "react";
...
+import defaultFeaturedItem from "../../data/defaultFeaturedItemData";
+
+import appConfig from "appConfig";
 
 const CampaignHero = ({ imageID }) => {
+  const defaultFeaturedItemResponse =
+    defaultFeaturedItem[appConfig["environment"]];
+  console.log("defaultItem is:", defaultFeaturedItemResponse);
+
   const [data, setData] = useState<any>({});
```
Comment (EdwinGuzman) [inline src/components/hero/campaignHero.tsx:16]: Here is another place where the `any` can be replaced with the `FeaturedItemData` interface.

**Review by EdwinGuzman — CHANGES_REQUESTED** [review]
The changes I am requesting are mostly for cleanup and to add the typescript interface. Otherwise, this works as intended so this is great!

One way I tested this is by adding `responseData.featuredItem.foregroundImageSrc = "foo.jpg";` in the `useEffect`:

```
const fetchData = async () => {
      const response = await fetch(`/api/featuredItem`);
      const responseData = await response.json();
      responseData.featuredItem.foregroundImageSrc = "foo.jpg";
      setData(responseData);
    };
```

This makes it so that "foo.jpg" will not load and the fallback image and text will load.

File: src/components/hero/campaignHero.tsx [hunk 4]
```diff
@@ -4,9 +4,19 @@ import { useEffect, useState } from "react";
 import CampaignHeroSubText from "./campaignHeroSubText";
 import CampaignHeroHeading from "./campaignHeroHeading";
 import CampaignHeroLoading from "./campaignHeroLoading";
+import defaultFeaturedItem from "../../data/defaultFeaturedItemData";
+
+import appConfig from "appConfig";
 
 const CampaignHero = ({ imageID }) => {
+  const defaultFeaturedItemResponse =
+    defaultFeaturedItem[appConfig["environment"]];
```
[the remark this answers is not in the document]
Comment (avertrees) [reply src/components/hero/campaignHero.tsx:13]: yes

File: src/utils/utils.ts [hunk 5]
```diff
@@ -26,13 +27,11 @@ export const imageURL = (
...
 export const getNumDigitizedItems = async () => {
   const apiUrl = `${process.env.API_URL}/api/v2/items/total`;
-  // console.log(`getNumDigitizedItems: About to fetch ${apiUrl}`);
-  // console.log(`getNumDigitizedItems calls apiCall function internally`);
   const res = await apiCall(apiUrl);
 
-  const fallbackCount = 863848;
-  const totalItems = res?.count?.$ || fallbackCount;
-  return addCommas(totalItems);
+  const fallbackCount = defaultFeaturedItems[appConfig.environment];
```
[the remark this answers is not in the document]
Comment (avertrees) [reply src/utils/utils.ts:32]: OOP good catch

File: src/components/hero/campaignHero.tsx [hunk 6]
```diff
@@ -4,9 +4,19 @@ import { useEffect, useState } from "react";
...
+  console.log("defaultItem is:", defaultFeaturedItemResponse);
+
   const [data, setData] = useState<any>({});
+  const [imageSrc, setImageSrc] = useState(
+    defaultFeaturedItemResponse.featuredItem.foregroundImageSrc
+  );
```
[the remark this answers is not in the document]
Comment (avertrees) [reply src/components/hero/campaignHero.tsx:19]: oop I forgot to remove that line before pushing.

File: src/components/hero/campaignHero.tsx [hunk 7]
```diff
@@ -18,6 +25,12 @@ const CampaignHero = ({ imageID }) => {
     fetchData();
   }, [imageID]);
 
+  const handleError = (e: any) => {
+    console.log(e);
+    setData(defaultFeaturedItemResponse);
+    console.log("data is:", data);
```
Comment (EdwinGuzman) [inline src/components/hero/campaignHero.tsx:31]: Nitpick: in the DS we add a prefix to know where the error is coming from. You can still see this in the stack trace but it's helpful if you just read the console message. So something like, "CampaignHero: the data is ...".

File: src/components/hero/campaignHero.tsx [hunk 8]
```diff
@@ -33,6 +46,10 @@ const CampaignHero = ({ imageID }) => {
       imageProps={{
         alt: data.featuredItem.title,
         src: data.featuredItem.foregroundImageSrc,
+        fallbackSrc: data.featuredItem.foregroundImageSrc,
```
Comment (EdwinGuzman) [inline src/components/hero/campaignHero.tsx:49]: Not necessary to have this anymore but it doesn't hurt to include.

**Review by EdwinGuzman — APPROVED** [review]
Some minor comments but works well.

**Review by 7emansell — APPROVED** [review]
Lgtm except as we talked about, I'd like to find a way to test this. I would expect something like this to work: 

```
it("renders the fallback image", async () => {
    (global as any).fetch = jest.fn(() =>
      Promise.reject(new Error("fails to fetch"))
    ) as jest.Mock;
    render(<CampaignHero imageID="5273165" />);
    await waitFor(() => {
      expect(
        screen.getByText("Watuppa, From water front, Brooklyn")
      ).toBeInTheDocument();
    });
  });
``` 
but it fails.

**Comment by EdwinGuzman** [comment]
Very true, I forgot about testing 😓 

Try `fireEvent.error` to trigger the `onerror` event and then check if the image has the fallback url. https://stackoverflow.com/questions/44462665/how-do-you-use-jest-to-test-img-onerror

**Pull request merged**